Add `useDataGrid` Hook Documentation and Example (#1830)
pankod/refine: documentation/docs/ui-frameworks/mui/hooks/useDataGrid.md

---
id: useDataGrid
title: useDataGrid
---

By using `useDataGrid`, you are able to get properties that are compatible with MUI X `<DataGrid>` component. All features such as sorting, filtering and pagination comes as out of box. For all the other features, you can refer to the MUI X [`<DataGrid>`][data-grid] documentation.

This hook is extended from [`useTable`](/core/hooks/useTable.md) from the [`@pankod/refine-core`](https://github.com/pankod/refine/tree/master/packages/core) package.

> 

Reviewer: Shall we add a `tip box` like "We support the Pro version" here?

## Basic usage

Let's assume that the data we are going to show on the table came like this from the endpoint:

```json title="https://api.fake-rest.refine.dev/posts"
[
    {
        "id": 1,
        "title": "A aspernatur rerum molestiae.",
        "content": "Natus molestias incidunt voluptatibus. Libero delectus facilis...",
        "status": "published"
    },
    {
        "id": 2,
        "title": "A molestiae vel voluptatem enim.",
        "content": "Voluptas consequatur quia beatae. Ipsa est qui culpa deleniti...",
        "status": "draft"
    }
]
```

To see basic usage, let's create a table with the `id`, `title` and `content` columns.

```tsx title="/src/pages/posts/list.tsx"

> import { useDataGrid, DataGrid, GridColumns, List } from "@pankod/refine-mui";

Reviewer: ```suggestion
import React from "react";
import { useDataGrid, DataGrid, GridColumns, List } from "@pankod/refine-mui";
```

    permanentSorter: [{ field: "title", order: "asc" }],
});
```

:::

:::tip

If you want to sort externally from the `<DataGrid>` component. You can use `setSorter` like this:

```tsx
import {
    useDataGrid,
    DataGrid,
    GridColumns,
    List,
    Button,
    ButtonGroup,
} from "@pankod/refine-mui";

const columns: GridColumns = [
    {
        field: "id",
        headerName: "ID",
        type: "number",
    },
    { field: "title", headerName: "Title" },
    { field: "status", headerName: "Status" },
];

export const PostsList: React.FC = () => {
    const { dataGridProps, setSorter } = useDataGrid({
        columns,
    });

    const handleSorting = (order: "asc" | "desc") => {
        setSorter([
            {
                field: "title",
                order,
            },
        ]);
    };

    return (
        <List>
            <ButtonGroup variant="outlined">
                <Button onClick={() => handleSorting("asc")}>Asc</Button>
                <Button onClick={() => handleSorting("desc")}>Desc</Button>
            </ButtonGroup>
            <DataGrid {...dataGridProps} autoHeight />
        </List>
    );
};
```

> Mui X community version only sort the rows according to one criterion at a time. To use multi-sorting, you need to upgrade to the [Pro plan](#).

Reviewer: ```suggestion
Mui X community version only sorts the rows according to one criterion at a time. To use multi-sorting, you need to upgrade to the [Pro plan](https://mui.com/pricing/).
```

```

When `sortModel` is not passed, it supports more than one criteria at a time, but cannot show which fields are sorted in `<DataGrid>` headers.

:::

## Filtering

The hook handles filtering by setting the `filterMode`, `filterModel` and `onFilterModelChange`props that are compatible with `<DataGrid>`.

```tsx
export const PostsList: React.FC = () => {
    const { dataGridProps } = useDataGrid({
        columns,
    });

    //highlight-start
    const {
        filterMode,
        filterModel,
        onFilterModelChange,
        ...restDataGridProps
    } = dataGridProps;
    //highlight-end

    return (
        <List>
            <DataGrid
                {...restDataGridProps}
                //highlight-start
                filterMode={filterMode}
                filterModel={filterModel}
                onFilterModelChange={onFilterModelChange}
                //highlight-end
                autoHeight
            />
        </List>
    );
};
```

Above, you can see the filtering properties from `dataGridProps`.

:::note
To see how the filtering works, you can look at the [source code][source-code] of the `useDataGrid` hook.
:::

> :::tip

Reviewer 2: Duplicated tip

However, multiple filtering can be done as server-side without specifying the `filterModel`.

```tsx
return <DataGrid {...dataGridProps} filterModel={undefined} autoHeight />;
```

When `filterModel` is not passed, it supports more than one criteria at a time, but cannot show which fields are filtered in `<DataGrid>` headers.

:::

## API

### Properties

| Key                                                          | Description                                                                                                                                                        | Type                                                                           | Default                                                                              |
| ------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------ |

> | columns<div className=" required">Required</div>             | Set of columns of type `GridColumns`                                                                                                                               | [`GridColDef`](https://mui.com/x/api/data-grid/grid-col-def)                   |                                                                                      |

Reviewer: ```suggestion
|  <div className="required-block"><div>columns</div> <div className=" required">Required</div></div>             | Set of columns of type `GridColumns`                                                                                                                               | [`GridColDef`](https://mui.com/x/api/data-grid/grid-col-def)                   |                                                                                      |
```

> | resource                                                     | The resource to use for table data                                                                                                                                 | `string` \| `undefined`                                                        | Resource name that it reads from the url                                             |

Reviewer: ```suggestion
| resource                                                     | The resource to use for table data                                                                                                                                 | `string` \| `undefined`                                                        | Resource name that it reads from the URL                                             |
```
